### DISCUSSION TITLE
Pretty output

### DISCUSSION THREAD

**Discussion by zavan** [yippee-fun/phlex#520] [Ideas]
It would be nice having the option to output pretty/indented HTML instead of the long string. I can use `Nokogiri` or `HtmlBeautifier` to do this, but they may change the output.

I think since Phlex already has the context of the opening/closing tags, maybe this is something that it could handle. Maybe as a plugin or a custom target?

**Comment by joeldrapper** [comment]
I’m all for this feature, but I think it would be quite difficult to implement it without impacting performance. I expect just checking a `pretty_output?` conditional once per tag will cost about ~1% in the case where we don't even want pretty output.

If this is a global setting, it might be possible to generate the tag methods with or without pretty output built-in. This wouldn't have any runtime performance impact on normal output, but you'd have to somehow have the setting configured before `Phlex::HTML::StandardElements`, etc. are required and the tag methods are built. Alternatively, there might be a way to trigger a rebuild of all the registered elements. You could probably do that by evaluating the contents of the files that register the elements but that wouldn't rebuild custom elements registered in other places.

**Comment by joelmoss** [comment]
This would be really nice actually, and I've found myself wanting it a few times. However, I see it as a dev only feature. So maybe it can be enabled globally in non-prod environments, as we don't care about performance as much there?

**Reply by zavan** [reply]
I would use it in production too. I'm using Phlex to output HTML into files and these files need to look nice. It doesn't have to be lightning fast for this, though.

**Reply by joeldrapper** [reply]
The problem is, just having the feature might make everything slower even if it’s disabled.
